Support overlapping Ivy patterns

ljacomet opened the pull request:
* Previously, only a single pattern could access a given location. We
now cache the listing results and return them for all patterns needing
it.
* When multiple patterns co-exist, we filter out results from a given
pattern that happen to match another one.

melix commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ResourceVersionLister.java:
```diff
@@ -144,17 +171,30 @@ private boolean revisionMatchesDirectoryName(String pattern) {
+
+    @Nullable
+    private List<String> listWithCache(ExternalResourceName parent) {
+        List<String> result;
```
I think you can simplify this with `computeIfAbsent`.

ljacomet commented:
@melix Simplified the code with `computeIfAbsent` and removed the need for a copy.

I also added a filtering of patterns to ignore ones that are identical or a subset of another one.

ljacomet commented:
Reverted the `computeIfAbsent` change because it will retry a compute when the key maps to `null`, which is not what we want here.
